pkg/controller/combo_controller.go

```diff
@@ -2,36 +2,161 @@ package controller
+	err := c.Get(ctx, req.NamespacedName, combination)
+	if err != nil {
 		return reconcile.Result{}, client.IgnoreNotFound(err)
 	}
 
-	return reconcile.Result{}, nil
+	// If combination is being deleted, remove from queue
+	if !combination.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("combination is being deleted, ignoring event")
+		return reconcile.Result{}, nil
+	}
+
+	// Attempt to retrieve the template referenced in the combination CR
+	template := &v1alpha1.Template{}
+	if err := c.Get(ctx, types.NamespacedName{Name: combination.Spec.Template}, template); err != nil {
+		combination.SetCondition(metav1.Condition{
+			Type:               v1alpha1.TypeInvalid,
+			Status:             metav1.ConditionFalse,
+			Reason:             v1alpha1.ReasonTemplateNotFound,
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Message:            fmt.Sprintf("failed to retrieve %s template: %s", combination.Spec.Template, err.Error()),
+		})
+		return reconcile.Result{}, errors.NewAggregate([]error{err, c.Status().Update(ctx, combination)})
+	}
+
+	// Build combination stream to be utilized in template builder
+	comboStream := combinationPkg.NewStream(
+		combinationPkg.WithArgs(formatArguments(combination.Spec.Arguments)),
+		combinationPkg.WithSolveAhead(),
+	)
+
+	// Create a new template builder
+	builder, err := templatePkg.NewBuilder(strings.NewReader(template.Spec.Body), comboStream)
+	if err != nil {
+		combination.SetCondition(metav1.Condition{
+			Type:               v1alpha1.TypeInvalid,
+			Status:             metav1.ConditionFalse,
+			Reason:             v1alpha1.ReasonTemplateBodyInvalid,
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Message:            fmt.Sprintf("failed to construct a builder out of %s template body:  %s", combination.Spec.Template, err.Error()),
+		})
+		return reconcile.Result{}, errors.NewAggregate([]error{err, c.Status().Update(ctx, combination)})
+	}
+
+	// Build the manifest combinations
+	generatedManifests, err := builder.Build(ctx)
+	if err != nil {
+		combination.SetCondition(metav1.Condition{
+			Type:               v1alpha1.TypeInvalid,
+			Status:             metav1.ConditionFalse,
+			Reason:             v1alpha1.ReasonEvaluationsInvalid,
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Message:            fmt.Sprintf("failed to generate manifest %s combinations: %s", combination.Spec.Template, err.Error()),
+		})
+		return reconcile.Result{}, errors.NewAggregate([]error{err, c.Status().Update(ctx, combination)})
+	}
+
+	log.Info(fmt.Sprintf("reconciliation of %s combination successful!", combination.Name))
+
+	// Add the combination evaluations and update Status
+	combination.Status.Evaluations = generatedManifests
+	combination.SetCondition(metav1.Condition{
+		Type:               v1alpha1.TypeInvalid,
+		Status:             metav1.ConditionTrue,
+		Reason:             v1alpha1.ReasonProcessed,
+		LastTransitionTime: metav1.NewTime(time.Now()),
```

Comment:
Be careful here. LastTransitionTime will change every time the object is reconciled, even if no real "work" was done. Instead, only update the object if the condition has changed, possibly using [this function](https://github.com/kubernetes/apimachinery/blob/e7b02e6514980d90b0df7364d3cb4c3316ace958/pkg/api/meta/conditions.go#L95) to check type/status.

Reply:
This is the only comment that I would classify as "possibly blocking".

Reply:
I believe that `combination.SetCondition` is already handling this logic. What are your thoughts around that function?

Reply:
Again, I would rather we use the upstream library whenever possible, but this does seem to meet the needs of my request.

Reply:
I agree, replaced this function with a thin wrapper around the one defined in `apimachinery`.